### ISSUE TITLE
Question: How to get the response body of an error?

### ISSUE THREAD

**Issue by ledikari** [Moya/Moya#1156]
Moya 8.0.0

```swift
func getMerchantDetails(qrId: String) -> Observable<RQRMerchant>? {
return Observable.create{ observer in
            let _ = self.provider
                .request(.getMerchantDetails(qrId: qrId))
                .filterSuccessfulStatusCodes()
                .mapJSON()
                .subscribe(onNext: { response in
                    observer.onNext(MyMerchant(json: JSON(response)))
                }, onError: { error in
                    print(error.localizedDescription)
                    observer.onError(error)
                })
            return Disposables.create()
            }
        }
}
```
assuming the api will return an error and I can retrieve the error code, how can i retrieve the response json payload?

**Comment by pedrovereza** [comment]
Hey @ledikari, this feature is planned to be release on Moya 9.0. See #993 :wink:

**Comment by sunshinejr** [comment]
Hey @ledikari. @pedrovereza is right, in 9.0.0 the `underlying` case would be getting response as well, but right now you can just cast error to `Moya.Error` and parse the response data:
```swift
.catchError { error in
    if let error = error as? Moya.Error, body = error.response?.mapJSON() {}
}
```

the syntax might be off since I'm writing this from the bus, but the idea should be alright. 😅

Let us know if it helped!

**Comment by ledikari** [comment]
@sunshinejr tried your code, but error.response is nil

`do {
                        let errorResponse = error as? Moya.MoyaError
                        if let body = try errorResponse?.response?.mapJSON(){
                            print(body)
                        }
                    } catch {
                        print(error)
                    }
`

**Comment by SD10** [comment]
@ledikari Do you know which case the `MoyaError` is? 
You can use `errorResponse?.errorDescription` if you'd like.

The reason I'm asking is the `MoyaError.underlying` and `MoyaError.requestMapping` return nil for the `response` property.

**Comment by ledikari** [comment]
@SD10 

> `You can use errorResponse?.errorDescription if you'd like.`
i tried printing the `errorResponse?.errorDescription` and the output is `Optional("Response status code was unacceptable: 470.")` how to get which case of MoyaError it is?

**Comment by SD10** [comment]
@ledikari It looks like you're receiving a `MoyaError.underlying`, so the response property will be `nil`.
You can view the output for `MoyaError.errorDescription` [here](https://github.com/Moya/Moya/blob/master/Sources/Moya/MoyaError.swift).

**Comment by ledikari** [comment]
@SD10  ohh i see, is there any workaround so that we could get the error response body without altering Moya?

**Comment by ledikari** [comment]
@SD10 ugh looks like i need to write alomofire again 😭

**Comment by sunshinejr** [comment]
Yes, we are sorry about that - just the `underlying` error & `body` needed in it was so uncommon we didn't really think about it till few months back. But we are closing in for the 9.0.0 version so please take a look at it once it is shipped. Sorry again for the inconvenience 😥

**Comment by jeryRazakarison** [comment]
@ledikari In my case, I use the 9.0.0-dev branch, and no problem for now, it is very stable. (And you have access to the error body:) )

**Comment by pedrovereza** [comment]
@ledikari @jeryRazakarison Yes, you can use the version of Moya that is under development if you'd like. I'd only suggest that you fork the repo and point to your own fork instead of relying on our `9.0.0-dev` branch as we can't guarantee it will *always* be stable :wink:

**Comment by ledikari** [comment]
@jeryRazakarison thats an interesting option. I will try to do it.

@sunshinejr any idea on when Moya 9.0 will be released?

**Comment by sunshinejr** [comment]
@ledikari Not sure, we are tracking progress in #1075, but slowly getting there. I didn't recommend using our `9.0.0-dev` since there might be some breaking changes in the near future, so please use it at your own risk 👍

**Comment by dimohamdy** [comment]
any update

**Comment by SD10** [comment]
Yes @dimohamdy this is available in Moya 9.0+
